server/src/main/java/org/elasticsearch/discovery/UnicastConfiguredHostsResolver.java

```diff
@@ -0,0 +1,105 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.discovery;
+
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.discovery.PeerFinder.ConfiguredHostsResolver;
+import org.elasticsearch.discovery.zen.UnicastHostsProvider;
+import org.elasticsearch.discovery.zen.UnicastZenPing;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public class UnicastConfiguredHostsResolver extends AbstractLifecycleComponent implements ConfiguredHostsResolver {
+
+    private final AtomicBoolean resolveInProgress = new AtomicBoolean();
+    private final TransportService transportService;
+    private final UnicastHostsProvider hostsProvider;
+    private final SetOnce<ExecutorService> executorService = new SetOnce<>();
+    private final TimeValue resolveTimeout;
+    private final Supplier<ExecutorService> executorServiceFactory;
+
+    public UnicastConfiguredHostsResolver(Settings settings, TransportService transportService, UnicastHostsProvider hostsProvider,
+                                          Supplier<ExecutorService> executorServiceFactory) {
+        super(settings);
+        this.transportService = transportService;
+        this.hostsProvider = hostsProvider;
+        resolveTimeout = UnicastZenPing.DISCOVERY_ZEN_PING_UNICAST_HOSTS_RESOLVE_TIMEOUT.get(settings);
+        this.executorServiceFactory = executorServiceFactory;
+    }
+
+    @Override
+    protected void doStart() {
+        executorService.set(executorServiceFactory.get());
+    }
+
+    @Override
+    protected void doStop() {
+        ThreadPool.terminate(executorService.get(), 10, TimeUnit.SECONDS);
+    }
+
+    @Override
+    protected void doClose() {
+    }
+
+    public void resolveConfiguredHosts(Consumer<List<TransportAddress>> consumer) {
+        assert lifecycle.started() : lifecycle;
```

Comment:
I wonder if this might be called while the service is being stopped. 
Note that we might also make use of `AbstractLifecycleRunnable` here (just found that class by chance)

Reply:
I changed the assert to a guard in a8a6393d45be92ec8a058ea973ad32d02c99326a but won't use `AbstractLifecycleRunnable` since it doesn't run `onAfterInLifecycle` method if the lifecycle state is `STOPPED`, which means this might break if stopped and then started again.

Reply:
However I also added a check for the lifecycle state in the action itself in c06a786782f5f018dc5c745f905c43a85062cc4e